Formula/bombadillo.rb

```diff
@@ -0,0 +1,23 @@
+class Bombadillo < Formula
+  desc "Non-web browser, designed for a growing list of protocols"
+  homepage "http://bombadillo.colorfield.space"
+  url "https://tildegit.org/sloum/bombadillo/archive/2.2.0.tar.gz"
+  sha256 "037a5e70b02fcf34b7d01f710111576b66faf117bceb000b527bff13ce939f3c"
+
+  depends_on "go" => :build
+
+  def install
+    system "go", "build", "-o", bin/"bombadillo"
+    man1.install "bombadillo.1"
+    prefix.install_metafiles
+  end
+
+  test do
+    require "pty"
+
+    PTY.spawn(bin/"bombadillo gopher://bombadillo.colorfield.space") do |r, w, _pid|
+      w.write "q"
+      assert_match /Bombadillo/, r.read
+    end
+  end
```

Comment:
You can avoid relying on network connection by using local gopher server stub:
```suggestion
  test do
    require "pty"

    begin
      port = free_port
      server = TCPServer.new(port)
      server_pid = fork do
        connection = server.accept
        msg = connection.gets
        assert_match /^\r?\n$/, msg
        connection.puts "iHello world!\tfalse\tnull.host\t1\r"
        connection.puts ".\r"
        connection.close
        server.close
      end

      cmd = "#{bin}/bombadillo gopher://127.0.0.1:#{port}"
      output, input, client_pid = PTY.spawn "sh -c 'stty rows 80 cols 24 && #{cmd}'"
      sleep 1
      input.putc "q"

      content = ""
      output.each do |line|
        next if line.start_with? "\e"

        content += line
      end
      assert_match "Hello world!", content

      Process.wait server_pid
    ensure
      Process.kill("TERM", client_pid)
    end
  end
```

Reply:
Please don't stub anything unless it's unavoidable

Reply:
You would rather depend on external service that could vanish anytime rendering test useless? 🤔

Reply:
Yes, because that would mean the program is also broken for users.

Reply:
Except program is _alt web_ browser (supporting gopher, gemini, and other alternative protocols) and browsers do not become broken if servers go down.

Reply:
So you pick a reliable host and test with that. We don't mock tests in homebrew formulae unless there's no other way.

Reply:
And that is consensus, right? https://github.com/Homebrew/homebrew-core/blob/8037f20de4b509c0f93277545619a1d2c8edc4b3/Formula/asuka.rb

Reply:
Full disclosure, I made that and nobody told me to avoid that stub. Needless to say, the whole point of gopher and gemini is the ease of implementation on both server and client-side. Declaring any host reliable in this area is very challenging by definition due to relatively low popularity and obscurity.

Reply:
We also don't generally accept version tests but sometimes there are exceptions or existing formula that violate this rule. Or maybe the person approving that link is human and made a mistake.

Reply:
For what it's worth, I assumed it was ok to use the external server based on https://github.com/Homebrew/homebrew-core/blob/master/Formula/wget.rb#L43.

Reply:
> For what it's worth, I assumed it was ok to use the external server based on [/Formula/wget.rb@`master`#L43](https://github.com/Homebrew/homebrew-core/blob/master/Formula/wget.rb#L43) [](/Homebrew/homebrew-core/blob/HEAD@{2020-05-18T16:04:03Z}/Formula/wget.rb#L43).

Well, authors gopherhole isn't exactly google but I'm kinda tired of making that point so let's leave that aside for a moment. However, the test still needs to be altered so you can actually verify that bombadillo loaded and rendered given gopher map.

```rb
  test do
    require "pty"

    begin
      cmd = "#{bin}/bombadillo gopher://bombadillo.colorfield.space"
      output, input, client_pid = PTY.spawn "sh -c 'stty rows 80 cols 43 && #{cmd}'"
      sleep 1
      input.putc "q"

      content = ""
      output.each do |line|
        next if line.start_with? "\e"

        content << line
      end
      assert_match /^\s+Bombadillo is a non-web browser/, content
      assert_match /^\(IMG\)\s+7\s+Bombadillo screen shot/, content
    ensure
      Process.kill("TERM", client_pid)
    end
```

The key part here is setting terminal size using `stty` because that is what bombadillo uses internally to calculate the output layout.

Reply:
It seems like this is covered by the test in https://github.com/Homebrew/homebrew-core/pull/54718/commits/7b1cba54acce2c5727bceb9bd2fa02d10901c9b6 -- it checks for some text on the page and sends the `q` keystroke to quit the program. But I agree that adding the `Process.kill` is a good idea. I'm also not sure whether we want that second assert, since that exact output might be subject to change.

Reply:
> I'm also not sure whether we want that second assert, since that exact output might be subject to change.

Now we are going full circle because anything out there in the wild could change and relying on external factors inside test is always a bad idea. At the same time, it seems it is also established practice so picking reasonably reliable and static Gemini space would be preferred. Maybe using gemini://gemini.circumlunar.space serves that purpose better:
https://github.com/Homebrew/homebrew-core/blob/c653db4/Formula/asuka.rb#L23-L40

To answer your question about the necessity of checking received/rendered content, well... This thing is a textual browser and the fact it can allocate screen and react upon quit instruction doesn't mean much to an end-user if the main aspect of network communication and content rendering is broken?